### PR TITLE
fix(amplify-category-function): use ref for S3Bucket and S3Key in CFN

### LINKED ISSUE #5387: Stop storing S3 bucket Lambda deployment information under Git
**Is your feature request related to a problem? Please describe.**

When you work in a team on a project with multiple Amplify envs, you merge code from multiple developers in a main branch (at some point).
The problem is that the Git history and the code reviews are polluted by changes done when switching environment instead of real code changes.
To be more specific, all the CloudFormation scripts under amplify/backend/functions/XXX contains section like this:

```
"Code": {	
	"S3Bucket": "blabla-env-123456-deployment",
	"S3Key": "amplify-builds/blabla-1234567890123456-build.zip"	
},
```

Because the S3Key and S3Bucket is different for all env (and the S3Key even change between push) these value change all the time creating a lot of fuzz in the Git commits.

**Describe the solution you'd like**

I like to have the Amplify CLI dynamically injecting that information when pushing and removing the hardcoded values from the CloudFormation files.
Like having:

```
"Code": {	
	"S3Bucket": "DO-NOT-CHANGE-UPDATED-BY-AMPLIFY-CLI",
	"S3Key": "DO-NOT-CHANGE-UPDATED-BY-AMPLIFY-CLI"	
},
```
And if the Amplify CLI needs to store the information about those URLs please use team-provider-info.json for information that remains and a file that is under .gitignore for the information that changes at every push.

### PR DESCRIPTION
**Pull request by jhockett**
*Issue #, if available:*
Closes #5387

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/amplify-provider-awscloudformation/src/push-resources.ts [hunk 1]
```diff
@@ -574,15 +560,26 @@ async function packageResource(context: $TSContext, resource: $TSAny) {
         Key: s3Key,
```
Comment (edwardfoyle) [inline packages/amplify-provider-awscloudformation/src/push-resources.ts:560]: should the cfn params also be set in this if block?

**Review by edwardfoyle — APPROVED** [review]
just one question, but LGTM

**Review by edwardfoyle — APPROVED** [review]

**Comment by r0zar** [comment]
:clap: :tada: thanks @jhockett

**Comment by Nxtra** [comment]
Does this mean I have to manually update my existing functions?

**Comment by jhockett** [comment]
> Does this mean I have to manually update my existing functions?

No, it should be taken care of automatically on the next `amplify push`. Did you run into any issues?

**Comment by Nxtra** [comment]
> > Does this mean I have to manually update my existing functions?
> 
> No, it should be taken care of automatically on the next `amplify push`. Did you run into any issues?

My existing functions weren't updated. But I now see, that you have to touch the function. Eg adding a console.log, there by having the function recognized as updated, will trigger the changes in the cloudformation template.

**Pull request merged**